Archive extensions

jedwards4b opened the pull request:
Extend archiving support to accept a regex pattern as well as a cesm style file extension.  This is 
intended to allow *any* format for history and restart file names.  Deserves a careful review. 

Edit from Bill Sacks: A model can define both extensions and regexes, but probably shouldn't (i.e., should probably just define one style or the other).

Test suite: scripts_regression_tests.py  (complete - all tests pass)
Test baseline: 
Test namelist changes: 
Test status: bit for bit

Fixes 

User interface changes?: 

Update gh-pages html (Y/N)?:

Code review:

jedwards4b commented:
@jgfouca this should allow us to get rid of the special mpas file handling code.

jedwards4b commented:
@uturuncoglu , @rsdunlapiv , @mvertens This PR will allow the SMS tests to work with fv3gfs.

jedwards4b commented:
@billsacks @jgfouca I am on leave for the holidays after tomorrow, I would really appreciate a review and if possible merge prior to that.   I am pretty sure that I have fixed all the problems but am in the process of rerunning scripts_regression_tests now.   Thanks

jgfouca commented on scripts/lib/CIME/hist_utils.py:
```diff
@@ -544,15 +509,16 @@ def _generate_baseline_impl(case, baseline_dir=None, allow_baseline_overwrite=Fa
-        else:
-            file_extensions = archive.get_hist_file_extensions(archive.get_entry(model))
-        hists =  _get_latest_hist_files(model, rundir, file_extensions, ref_case=ref_case)
+#        if model == 'cpl':
```
Remove commented-out code

jedwards4b replied:
done

jgfouca commented:
How good is the coverage of this stuff in scipts-regression-tests?

billsacks commented:
@jedwards4b I'm in the midst of some time-critical work. I can review this as soon as that is complete, but that may not happen before end of the day tomorrow (it depends on how quickly I can resolve some issues I'm having).

jedwards4b commented:
@billsacks that is unfortunate as I feel you are best qualified to review.  Would you be willing to allow the merge to master to go forward and address any issues you find later?

billsacks commented:
> Would you be willing to allow the merge to master to go forward and address any issues you find later?

Yes.

billsacks commented:
>> Would you be willing to allow the merge to master to go forward and address any issues you find later?
>
> Yes.

@jedwards4b thinking about this more: In the past, it has been more painful to do reviews and get changes made after the fact. I imagine you're asking for this because some people need this change? If so, is it feasible for them to point to this branch for the next couple of weeks? If that would be problematic for some reason, I'm still willing to let this come to master without a review.

jedwards4b commented:
I will leave it on the branch - please review and (if appropriate) merge when you have a chance.

billsacks commented:
@jedwards4b I just talked with @mvertens based on your chat messages with her. Just to reiterate: If there is a big benefit to having this on master ASAP, then I'm truly okay with that.

jedwards4b commented:
@billsacks The UFS developers need this for their work but should be able to use the branch until you are ready to merge.

billsacks commented on scripts/lib/CIME/case/case_st_archive.py:
```diff
@@ -257,36 +256,22 @@ def _archive_history_files(archive, archive_entry,
-
-        pfile = re.compile(newsuffix)
-        histfiles = [f for f in os.listdir(rundir) if pfile.search(f)]
-        logger.debug("histfiles = {} ".format(histfiles))
```
I haven't taken the time to understand this deletion. Given your request for a careful review: Do you want a careful review of this change, or are you pretty confident about this one?

jedwards4b replied:
I have tested case.st_archive --test-all with CIME_MODEL=e3sm and all pass.

billsacks commented on scripts/lib/CIME/case/case_st_archive.py:
```diff
@@ -652,22 +637,24 @@ def restore_from_archive(self, rest_dir=None, dout_s_root=None, rundir=None):
+        rest_root = os.path.join(dout_s_root, "rest")
+        if os.path.exists(rest_root):
+            rest_dir = os.path.join(rest_root, ls_sorted_by_mtime(os.path.join(dout_s_root, "rest"))[-1])
+    if rest_dir:
```
Can you explain this change? It looks like it is allowing for the possibility that the restart directory doesn't exist... but naively, I would have thought that that should be an error.

jedwards4b replied:
I don't think that there is a change here, just reordered some code.

billsacks replied:
I mean adding the conditional `if rest_dir` around the block of code below. You have changed some of the above code so that there is now a possibility that `rest_dir` won't be set: if you enter the `else` block just above this, but then `if os.path.exists(rest_root)` is False. My sense is that this function should abort with an error message in the case that `rest_dir` remains unset at this point, rather than silently doing nothing (which is the case with your changes, I think). If there's a reason that it's okay for `rest_dir` to remain unset and for this to end up doing no copies, can you describe what that scenario would be?

jedwards4b commented on scripts/lib/CIME/XML/archive_base.py:
```diff
@@ -3,6 +3,7 @@
 """
 from CIME.XML.standard_module_setup import *
 from CIME.XML.generic_xml import GenericXML
+from CIME.hist_utils import get_extension
```
done

jedwards4b commented:
@billsacks I think that I've addressed your concerns.  I fixed a problem with recognizing the cime_model in an exisiting case.   I fixed an issue with finding the config_archive.xml file used for e3sm and made sure that mpas and mali files passed the test.   I removed a special case for history files with initial in the extension since that seems to contradict a test for cism archives.

billsacks commented:
@jedwards4b thank you for your changes addressing some of my requests. Most of your latest changes look good. However, I think the removal of the special handling of 'initial' is a problem: If I remember correctly, it causes problems if CISM's 'initial_hist' file is found in ER* tests, since this file is present in the initial case but not the restart case. (See some related discussion in #2728.) As per the discussion there, this file should be archived, but should NOT be examined by the test system. (I wonder if this will now require some more general implementation like what I suggested in https://github.com/esmci/cime/pull/2728#issuecomment-407568275 ?)

However, there are still a number of review points that have not been addressed:

- Requests in my overall comment still need to be addressed (https://github.com/ESMCI/cime/pull/3326#pullrequestreview-336613487)

- Comments not addressed (some of these were "helpfully" hidden by GitHub... you may need to click on "load more..." under "4 hidden conversations")
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361531673
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361547834
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361548311
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361548793
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361549637 (okay to just say that that is too much for now)
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361551476 (I apologize about this one: it is an amendment to an earlier comment, and I'm sorry if the way I did this was confusing)
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361752291
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361753295 (just let me know if you want a careful review of this set of changes)
	- https://github.com/ESMCI/cime/pull/3326#discussion_r361753746

jedwards4b commented:
@billsacks - I have moved the special handler for initial hist files to system_test_common.py

jedwards4b commented:
> Is the expectation that a given model will define either extensions or regex? Or will it work fine for a given model to define both? I think the latter is okay, but can you confirm?

A model can define both, but probably shouldn't.

jedwards4b commented:
I'm not going to address the larger issues about what python and xml files should be named or whether a given model should be allowed to have stricter naming conventions.

jedwards4b commented on scripts/lib/CIME/hist_utils.py:
```diff
@@ -183,7 +145,8 @@ def _hists_match(model, hists1, hists2, suffix1="", suffix2=""):
     for hists, suffix, normalized, multi_normalized in [(hists1, suffix1, normalized1, multi_normalized1), (hists2, suffix2, normalized2, multi_normalized2)]:
         for hist in hists:
-            normalized_name = hist[hist.rfind(model):]
+            offset = max(0,hist.rfind(model))
```
I have generalized so that history files do not need to include the model name.  I'm sure it's not 100% fool proof.  If the normalized filename does something like leave the date out then either we will need to improve the scripts or the user / component developer could consider improving their file naming convention.  @mvertens do we currently have documented the history file naming convention for cime?

billsacks replied:
Here are the current file name conventions for CESM: http://www.cesm.ucar.edu/models/cesm2/naming_conventions.html#outputFilenames

I feel fairly strongly that, if we are going to generalize these, then we need to clearly document what we do (and possibly don't) expect to work. Patterns that are expected to work should be verified via unit tests of the relevant functions. Saying something to the effect of, "We kind of support any file names, but some things may not work if you depart too far from our existing files" (I'm paraphrasing here: I know you didn't say that directly) is confusing for component model developers and makes it very hard for cime developers to know what they are supposed to support in the scripts. While the ideal case would be to have a less restrictive file naming convention that is both clearly documented and well tested, I'd rather have something restrictive but clearly documented than something less restrictive but vague and uncertain.

I sense that an in-person discussion may be needed here. We could discuss this at the cime telecon Jan 8, but it might also be good if you and I talk along with @mvertens this week.

jedwards4b commented on scripts/lib/CIME/XML/archive_base.py:
```diff
@@ -12,23 +13,110 @@ def get_entry(self, compname):
+            histlist.append(os.path.join(from_dir,latest_files[key]))
+            # Special case for fv3gfs which outputs in cubed sphere tiles
+            if "tile[1-6].nc" in key:
+                for i in range(1,5):
```
I do mean 1-5, tile6 is already in histlist.

billsacks replied:
But `range(1,5)` doesn't even include 5:

```python
list(range(1,5))
[1, 2, 3, 4]
```

jedwards4b commented on scripts/lib/CIME/XML/archive_base.py:
```diff
@@ -12,23 +13,110 @@ def get_entry(self, compname):
+
+        for key in latest_files.keys():
+            histlist.append(os.path.join(from_dir,latest_files[key]))
+            # Special case for fv3gfs which outputs in cubed sphere tiles
```
removed hard coded behavior.

billsacks commented:
> @billsacks - I have moved the special handler for initial hist files to system_test_common.py

Excellent, thanks - I like your solution to that: it seems better than what was there originally.

jedwards4b commented:
@billsacks @jgfouca please revisit this review, I have scaled back the scope of changes and removed the regex feature so that the original description of cime filename conventions is still valid.

jedwards4b commented:
@billsacks after our meeting last week I decided to update this PR rather than to start over.   It's once again ready for your review.

jedwards4b commented:
@billsacks I think I've addressed your comments.   Thanks

billsacks commented:
@jedwards4b thank you for addressing most of my comments. There are still two outstanding points from my overall comment (https://github.com/ESMCI/cime/pull/3326#pullrequestreview-339360646). If you feel they shouldn't be changed, feel free to say so; I just want to be sure they aren't overlooked.

jedwards4b commented:
@billsacks I think I've covered them now

jedwards4b commented on scripts/lib/CIME/XML/archive_base.py:
```diff
@@ -12,23 +13,110 @@ def get_entry(self, compname):
+        for ext in extensions:
+            if 'initial' in ext:
+                continue
+            if ext.endswith('$') and len(suffix)>0:
```
I have tested against e3sm, cesm and ufs - I believe that this change is correct and found that your suggestion broke some history files like casename.cam_0002.e.preassim.1976-01-01-00000.nc

billsacks replied:
Okay, thanks for confirming that this change is intentional. Yes, I see that, for example, CLM defines `<hist_file_extension>e</hist_file_extension>`, so it's intentional that some other stuff can come after this.

billsacks commented:
Okay, looks good to me now. Thanks a lot for all of your work on this!
